Devx-1685: Refactoring prep-work

MonicaG opened the pull request:
This is prep-work for Devx-1685: Create "register your repo in dev hub" wizard.

The task is to create a wizard that will add the catalog-info.yaml, license and pubcode.yaml files to an existing repo. Since these are files that are used by multiple wizards, I wanted to place the code in a shared location. This is so I wasn't copy/pasting the parameters/actions into the new wizard.

This PR is to put those steps and actions in a shared location. Once the code is merge to main, I will create a separate PR for the wizard. The wizard will use these code changes.

Somethings I would like comments on (although please comment on other things as well):
- Is the `developer.gov.bc.ca/source_template:`  needed?
  - I removed that  field from the `catalog-info.yaml` file, as there is already a `backstage.io/source-template:` field. They contain duplicate info. Please let me know if both fields are needed.
- Suggestions for a name other than `template-snippets` folder. 
  - I put the shared code into a new folder called `template-snippets` folder in the `shared` folder. I'm not sure if I like that name. Open to suggestions on another name.
- `catalog-info-action.yaml` contains calls to `parameter.owner` and `parameter.repo` that are not in the `catalog-info-parameters.yaml` file. 
  - In my case I have an initial screen where the user provides the repo and owner details, and then a second screen where they include info about the title, description etc. I was hesitant to put the repo in the `catalog-info-parameters.yaml` file because I could see things needing to be specific to a wizard such as `allowedOwners` and `additionalScopes`. Thoughts on this?
  - Maybe I put comment at top of yaml snippet stating what is expected?
  - Do we want to have both shared parameters and actions? Or should it just be parameters that are shared and everyone writes their own actions? 

The snippets shown here can be used in a template if you want to try them out. For example:

```
parameters:
    - title: Initial form
       properties: <...snipped...>
    - $yaml: https://github.com/bcgov/devhub-templates/blob/DEVX-1685/shared/template-snippets/catalog-info-parameters.yaml
    - $yaml: https://github.com/bcgov/devhub-templates/blob/DEVX-1685/shared/template-snippets/pubcode-parameters.yaml
    - $yaml: https://github.com/bcgov/devhub-templates/blob/DEVX-1685/shared/template-snippets/license-parameters.yaml

steps:
    - id: some step
      name: <...snipped...>
    - $yaml: https://github.com/bcgov/devhub-templates/blob/DEVX-1685/shared/template-snippets/catalog-info-action.yaml
    - $yaml: https://github.com/bcgov/devhub-templates/blob/DEVX-1685/shared/template-snippets/pubcode-action.yaml
    - $yaml: https://github.com/bcgov/devhub-templates/blob/DEVX-1685/shared/template-snippets/license-action.yaml

``` 

These snippets use this branch. For the new wizard I'm creating I will use these snippets, but located on the main branch rather than this branch.

I used a local template to create a pull-request on [mlg-test-repo-1](https://github.com/bcgov/mlg-test-repo-1/pulls). The local template uses the snippets in this PR.

MonicaG committed:
Refactored out common parameters

MonicaG committed:
Refactore catalog info parameters into snippets

MonicaG committed:
Renamed directory and added shared action steps

MonicaG committed:
Use URL rather than relative path

MonicaG committed:
trying with same SHA

MonicaG committed:
refactor

MonicaG committed:
try with filename

MonicaG committed:
refactor

MonicaG commented:
@jenreiher-bcgov - For the new wizard, should it include license details too, or just the catalog-info and pubcode yaml files? Thanks!

jenreiher-bcgov commented:
> @jenreiher-bcgov - For the new wizard, should it include license details too, or just the catalog-info and pubcode yaml files? Thanks!

good question @MonicaG  - I can't think of any reason why we WOULDN'T want to include the license, can you?

MonicaG committed:
Add license parameters and actions

jenreiher-bcgov commented:
Not sure if this is helpful or not, but I think we were discussing NOT replicating the pub code as a separate file -even NR was questioning that

MonicaG commented:
> Not sure if this is helpful or not, but I think we were discussing NOT replicating the pub code as a separate file -even NR was questioning that

@jenreiher-bcgov - how should pubcode be represented? Is it needed?

MonicaG commented:
> > @jenreiher-bcgov - For the new wizard, should it include license details too, or just the catalog-info and pubcode yaml files? Thanks!
> 
> good question @MonicaG - I can't think of any reason why we WOULDN'T want to include the license, can you?

I've update this PR to include license as well.

MonicaG commented on shared/template-snippets/license-action.yaml:
```diff
@@ -0,0 +1,7 @@
+if: ${{ parameters.licenseType !== 'None' }}
+action: fetch:plain:file
+input:
+    url: https://github.com/bcgov/devhub-templates/tree/main/shared/licenses/${{ parameters.licenseType }}/LICENSE
```
Here and in the pubcode-action.yaml file, I changed the url to use a fixed location rather than relative location. The relative location will be relative to the template file that uses this license-action.yaml (or pubcode-action.yaml) file. Right now all our templates are in `software-templates/template-name` so the relative `../../shared/licenses/....` is valid. But I don't know if that would always be the case, so changed it to the fixed location. Happy to change it to relative if we think templates will always be in `software-templates/template-name`  with `shared` folder 2 directories about it.

oomIRL reviewed (approved):
Looks great. I'm keen to continue this refactor effort!

- **Is the developer.gov.bc.ca/source_template: needed?**
Just using `backstage.io/source-template` makes sense to me.

- **Suggestions for a name other than template-snippets folder.**
common-yaml? I'm good with whatever you choose

- **catalog-info-action.yaml contains calls to parameter.owner and parameter.repo that are not in the catalog-info-parameters.yaml file.**
I've been thinking about this as well, I think a comment at the top indicating dependencies is good. We can't always completely isolate each piece so documenting requirements is the best step imo.
I like the idea of having `parameters` and `actions`. Currently, the license and metadata actions seem quite resuable, but I expect others (like publish) won't be generic enough to be reusable.

MonicaG committed:
Updates from PR comments

MonicaG committed:
remove pubcode as per PR comment

MonicaG committed:
Removed email address from input template, was not used in output

MonicaG committed:
Added product description to bcgovpubcode

MonicaG committed:
Always provide owner field even when value is empty

MonicaG committed:
Add comment

MonicaG commented:
@jenreiher-bcgov @oomIRL - thank you for your comments! I've made the following changes based on your feedback:

- pubcode - removed the `pubcode/pubcode.yaml` and `pubcode-action.yaml` files. 
  - Pubcode is now written only to the catalog-info.yaml file
- renamed `template-snippets` to `common-yaml`
- Removed product owner email from `pubcode-parameters.yaml` as it is not used in the currently [pubcode.yaml](https://github.com/bcgov/devhub-templates/blob/main/shared/pubcode/pubcode.yaml) file in the main branch. 
- Renamed `catalog-info-action.yaml` to `catalog-info-component-action.yaml` to reflect that this action is for the [Component Kind](https://backstage.io/docs/features/software-catalog/descriptor-format/#kind-component). Different kind's of catalog-info files can be created based on the kind of Backstage entity
- Renamed `catalog-info-parameters.yaml` to `catalog-info-component-parameters.yaml` for the same reason as above
- No longer defaults the owner to our team if a team is not provided. 
  - I'm not sure about this. Backstage says [owner is required](https://backstage.io/docs/features/software-catalog/descriptor-format/#specowner-required), but I don't think we want it defaulting to our team if one is not provided. Maybe we require that owner field is required in the inputs?
- Added comments to the top of the action/parameter files 

You can view a PR based on these changes in the [mlg-test-repo-1](https://github.com/bcgov/mlg-test-repo-1/pulls) repo

jenreiher-bcgov commented:
@MonicaG for default team, I can see a few things - like we could have a "code-with-no-owner" team that's the default team  and that helps folks know that it's something that "should be changed"

or there's also a thought I've been mulling over with galen in the wizards context of there being an explicit team for code that is owned and maintained by the community (vs a team) as a signal of things that are "open for a different type of contribution" and we could consider making this the default team...?

MonicaG commented:
> @MonicaG for default team, I can see a few things - like we could have a "code-with-no-owner" team that's the default team and that helps folks know that it's something that "should be changed"
> 
> or there's also a thought I've been mulling over with galen in the wizards context of there being an explicit team for code that is owned and maintained by the community (vs a team) as a signal of things that are "open for a different type of contribution" and we could consider making this the default team...?

@jenreiher-bcgov - What are your thoughts on making the team field required?

MonicaG commented:
> > @MonicaG for default team, I can see a few things - like we could have a "code-with-no-owner" team that's the default team and that helps folks know that it's something that "should be changed"
> > or there's also a thought I've been mulling over with galen in the wizards context of there being an explicit team for code that is owned and maintained by the community (vs a team) as a signal of things that are "open for a different type of contribution" and we could consider making this the default team...?
> 
> @jenreiher-bcgov - What are your thoughts on making the team field required?

@jenreiher-bcgov - Actually, I guess regardless of whether we make it required or not, we should still have a fall back incase one template doesn't have it mandatory. I'll make a "devhub-wizards-default-team" team. I'll default to that if owner isn't set. Then in the "next steps" phase if that team is set, a template could provide a message that a team needs to be set on a repo.

I like the idea of the community team, but I don't want to set it here as default. As I see that as a separate use case.

Ok?

jenreiher-bcgov commented:
@MonicaG I think there was some blockers for making teams a required field that came up in planning yesterday. we would need to action the ticket around making a teams picker widget, and then have it only pull in teams that you have access to by default. This is not currently planned for this quarter, and sounded like a medium lift. If it's a blocker for this work we can definitely re-prioritize!

devhub-wizards-default-team sounds good to me :) love the self-documenting naming ;) we might want to have some tickets to monitor how many repos get created with this to see if folks are actually taking the action we want to take e.g. picking an appropriate team

MonicaG committed:
Default team to 'devhub-wizards-default-team' in catalog-info

MonicaG committed:
Update team to include group

MonicaG committed:
parameterize group value

MonicaG committed:
format default owner
